Add offline support

fjvallarino opened the pull request:
## Description
This adds support for offline play, using cached data. In case a fetch request fails and we detect the client is offline, we will try to deliver the latest cached response (ignoring expiration date).

To detect if the machine is offline we rely on https://widgets.risevision.com, which should be available in the users machine and has appropriate CORS settings.

## Motivation and Context
Rise Image and Rise Video handle cached content using Player provided features. These are not available, as far as I understand, for content not hosted on GCS. Rise RSS needs to support offline play.

## How Has This Been Tested?
Testing has been performed using Chrome Dev Tools to simulate offline scenarios, while using the Demo page for `rise-data-rss`. More specifically, a breakpoint was set before performing the first `fetch` call and only then the Offline mode was set (otherwise Chrome does not load the Demo page, which is served through a local web server).

## Release Plan:
- As the Submitter, upon requesting review of this pull request, I confirm that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed. 
- As the Reviewer, upon approving the changes in this PR, I confirm I have reviewed and I agree that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed

#### Release Checklist Items Skipped?
No. This changeset will only impact components which specifically use the new tag.

----

I thought about creating a completely separate mixin, but there were many features I needed from the existing `CacheMixin` and `FetchMixin` mixins; adding two new configuration flags made more sense.

@santiagonoguez @stulees please review

@alex-deaconu you are the main contributor to these two mixins. Can I bother you to take a look? Thanks!

alex-deaconu commented on src/cache-mixin.js:
```diff
@@ -5,6 +5,7 @@ export const CacheMixin = dedupingMixin( base => {
   const CACHE_CONFIG = {
       name: "cache-mixin",
       duration: 1000 * 60 * 60 * 2,
+      preserveExpired: false
```
@fjvallarino having written both these mixins, the way I would have implemented it is by adding a second `expiration` timer. Basically, duration is how long cache is valid (ie. we don't hit the API again) and expiration is how long it's kept for offline (before it gets removed). Setting any of the timers to -1 ignores that functionality and either hits the API every time (except for offline) or uses the offline version all the time (until it expires).

Anyways, that is just my suggestion for implementation. Adding this flag is OK too but I'm going to need to re-write once we implement the above behaviour. To confirm, for weather we are looking to improve and retrieve data every 30 minutes, but have an expiration of 2 hours. I believe this can be achieved with what I described above, but not with the `preserveExpired` flag.

I have a prototype branch for this if you want to take a look.

@ezequielc FYI.

Thanks.

alex-deaconu replied:
This is the implementation I mentioned: https://github.com/Rise-Vision/rise-common-component/compare/feature/fresh-data?expand=1. Thanks.

fjvallarino replied:
@alex-deaconu that makes sense and is more flexible, although in our use case we don't want the content to expire (of course, we could set a really high value for the new `expiration` timer).

I'm fine with either option. Considering we're way behind schedule on our epic I'm leaning on keeping this solution and updating `rise-data-rss` when the `rise-data-weather`/`rise-data-component` changes are in. Any component relying on this implementation (well, just 
`rise-data-rss`) will only need to update the `initCache` call and a few unit tests to upgrade. Let me know if you feel strongly about going to other route.

As a lesson learned, I guess this kind of changes need to be brought up for discussion before implementing them (just in case other team needs something similar but not quite).

stulees replied:
@fjvallarino Both approaches look good but it does seem what  @alex-deaconu implemented can cover both use cases better, so given _rise-data-rss_ is still in development and not used by anything, I think now would be the best time to refactor to work with `expiration` timer. Thoughts?

fjvallarino replied:
@alex-deaconu @stulees updated to account for the suggested use case!

fjvallarino commented:
@alex-deaconu @santiagonoguez @stulees updated!

santiagonoguez commented on README.md:
```diff
@@ -92,7 +92,14 @@ For caching arbitrary data responses, the mixin uses browsers Cache API.
 Whenever the cached data is retrieved, the mixin checks the date header and delete it from cache in case it is expired. Also, to prevent cache from growing indefinitely, during mixin initialization all expired cache entries are deleted.
 
-To enable the mixin in your component you have to declare the mixin and call the `initCache` function with your desired cache name and optionally duration. Cache duration defaults to 2h. 
+To enable the mixin in your component you have to declare the mixin and call the `initCache` function with your desired cache name and optionally duration. Cache duration defaults to 2h. Cache expiry defaults to 4h. Caches entries are removed using the expiry attribute; if set to -1, entries are not removed. Default values are:
```
@fjvallarino I would improve explanation of difference between duration and expiry, because it may be confusing. Alex explanation in his comment was clear, so we may incorporate that text here, thoughts ?

fjvallarino replied:
@santiagonoguez updated!

santiagonoguez commented on src/cache-mixin.js:
```diff
@@ -62,23 +71,18 @@ export const CacheMixin = dedupingMixin( base => {
+        if ( !this._isResponseExpired( response, this.cacheConfig.duration )) {
+          return Promise.resolve( response );
+        } else if ( !this._isResponseExpired( response, this.cacheConfig.expiry )) {
+          return Promise.reject( response );
```
@fjvallarino I'm still a little lost here :'( I don't understand why the response is rejected if the response hasn't expired according to the conditional, thanks ! :D

stulees replied:
@santiagonoguez Have a look at the `catch` of the `_getData()` function in fetch mixin - https://github.com/Rise-Vision/rise-common-component/blob/41e73795683658baafa27c72b7e53a971c043645/src/fetch-mixin.js#L66-L69

fjvallarino replied:
@santiagonoguez the branch that is rejected is when the object is stale (it's older than what `duration` allows) but it's still valid in terms of offline usage (younger than `expiry` or `expiry === -1`). In the case it's valid for offline usage, we don't just provide the cached data, but we also do some error handling and logging (in the section Stu pointed to).

santiagonoguez replied:
@stulees @fjvallarino I see :O I find the flow a little weird, but it's ok, thanks !

alex-deaconu commented on src/fetch-mixin.js:
```diff
@@ -102,9 +121,13 @@ export const FetchMixin = dedupingMixin( base => {
         this._requestRetryCount = 0;
 
-        super.log( "error", "request error", { error: err ? err.message : null });
+        if ( err && err.isOffline ) {
```
@fjvallarino is this the RSS parser specific error you were mentioning? Thanks.

fjvallarino replied:
@alex-deaconu it's not really `rise-data-rss` specific. I needed the `isOffline` flag to do some things in that component (there is a separate PR in that project, if you're interested), but I think it's useful for other components to log different messages when a request failed or when we're offline.

alex-deaconu replied:
@fjvallarino makes sense, thanks!

alex-deaconu commented on src/fetch-mixin.js:
```diff
@@ -78,7 +80,24 @@ export const FetchMixin = dedupingMixin( base => {
-      }).catch( this._handleFetchError.bind( this ));
+      }).catch( err => {
+        return this._isOffline().then( isOffline => {
+          cachedResp && this._processData( Object.assign( cachedResp, { isOffline }));
```
@fjvallarino wondering if there's a reliable way to determine if the display is offline from the fetch response. I know usually the response code is `-1` for that. Thanks.

fjvallarino replied:
@alex-deaconu when `fetch` rejects it's because there were network issues. It could be the destination server was down, or the client has network issues. That's why I added the extra call to a different resource: if that one fails, there are better chances that the issue is on the client itself.

alex-deaconu replied:
@fjvallarino actually, take a look at this: https://www.tjvantoll.com/2015/09/13/fetch-and-errors/. To confirm, reject means offline, while `!resp.ok` is a server error. I don't think we need the extra check. Thanks.

alex-deaconu replied:
@fjvallarino got it, not ideal but if that's the requirement than I guess it makes sense. Thanks.

fjvallarino replied:
@alex-deaconu yes, it's a bit unclear. Mozilla documentation mentions this (https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch): 

_The Promise returned from fetch() won’t reject on HTTP error status even if the response is an HTTP 404 or 500. Instead, it will resolve normally (with ok status set to false), **and it will only reject on network failure or if anything prevented the request from completing.**_

That's what I considered when adding the extra call.

alex-deaconu commented:
@fjvallarino the branch looks good. However, I think you need to add some test coverage over the two different timers in Cache. 

Also, this came up when @ezequielc evaluated my branch originally, is the fact that fetch returns data and throws an error when offline. This means it will send cached data 5 times (for each of the retries) and then throw an error, which may be confusing to the component/template. I think we need to send cached data only once, and your flags would address the error.

Thanks.

fjvallarino commented:
@alex-deaconu added/modified unit tests in `CacheMixin` to account for the different types of expiration parameters. I think it covers what we need, but let me know if you'd like to add something else.

Also added logic to only provide cached data the first time after an unsuccessful fetch, with the corresponding unit tests. I modified the counting logic in `_handleFetchError` a bit to account for this (it used to reset the counter after every five retriers). Let me know if you prefer to provide the cached data the first round of every block of five attempts (I put the counting changes in the latest commit, so it's easy to revert)

alex-deaconu commented on src/fetch-mixin.js:
```diff
@@ -136,6 +136,9 @@ export const FetchMixin = dedupingMixin( base => {
       }
     }
 
+    _shouldRetryWithErrorHandling() {
```
@fjvallarino I find this function name confusing. I would re-name to something like `!_requireCooldown()` or `_isMaxRetryAttempt()` (depending on whether you want to return true or false). Thanks.

alex-deaconu commented:
@fjvallarino despite my inline comment, I don't believe the last 2 commits are necessary. I think returning cached data once per 5 retries is ok, even if it's returned every time the 5 retries are performed. Thanks.

fjvallarino commented:
@alex-deaconu just to be clear, of the latest three commits (starting on 9f890e8):

- We want to keep the logic for only providing data on the first of every five retries
- We want to keep the unit tests that were added
- We want to keep that function renaming it to `_requireCooldown` or similar (I still need for one of my tests in `rise-data-rss`)
- The rest can be reverted.

Is that correct?

alex-deaconu commented:
@fjvallarino yes, sorry I was not being more specific. In my opinion the functionality in the `03e5b4e08` commit is overkill. If you need the extra utility function for your tests that's not an issue. Thanks.

fjvallarino commented:
@alex-deaconu updated!
